main/lsp/LSPTypechecker.cc

```diff
@@ -390,141 +331,214 @@ LSPTypechecker::SlowPathResult LSPTypechecker::runSlowPath(LSPFileUpdates &updat
+                finalGS->errorQueue,
+                make_shared<core::ErrorQueue>(finalGS->errorQueue->logger, finalGS->errorQueue->tracer, errorFlusher));
 
-                ShowOperation op(*config, ShowOperation::Kind::Indexing);
+            std::optional<Timer> timeit;
+            ShowOperation op(*config, ShowOperation::Kind::Indexing);
 
-                std::unique_ptr<const OwnedKeyValueStore> ownedKvstore =
-                    cache::ownIfUnchanged(*finalGS, std::move(kvstore));
+            switch (mode) {
+                // Initialization fetches the list of files to index from the options
+                case SlowPathMode::Init: {
+                    ENFORCE(!this->initialized);
+                    timeit.emplace(this->config->logger, "initial_index");
 
-                {
-                    Timer timeit(config->logger, "reIndexFromFileSystem");
-
-                    auto inputFiles = pipeline::reserveFiles(*finalGS, config->opts.inputFileNames);
-                    this->indexed.clear();
-                    this->indexed.resize(finalGS->filesUsed());
-
-                    auto asts = hashing::Hashing::indexAndComputeFileHashes(*finalGS, config->opts, *config->logger,
-                                                                            absl::Span<core::FileRef>(inputFiles),
-                                                                            workers, ownedKvstore);
-                    ENFORCE(asts.hasResult(), "LSP initialization does not support cancellation");
-                    // asts are in fref order, but we (currently) don't index and compute file hashes for payload files,
-                    // so vector index != FileRef ID. Fix that by slotting them into `indexed`.
-                    for (auto &ast : asts.result()) {
-                        int id = ast.file.id();
-                        ENFORCE_NO_TIMER(id < this->indexed.size());
-                        this->indexed[id] = std::move(ast);
+                    this->workspaceFiles = pipeline::reserveFiles(*finalGS, config->opts.inputFileNames);
+                    break;
+                }
+
+                // Reindexing on the slow path derives the list of inputs files from the file table
+                case SlowPathMode::Cancelable: {
+                    timeit.emplace(this->config->logger, "slow_path_reindex");
+
+                    ENFORCE(updates.updatedFiles.size() == updates.updatedFileIndexes.size());
+
+                    // Move the indexed trees into our local cache for open files. We will still reindex them again to
+                    // force any errors, but this is a convenient way to pre-populate our cache.
+                    if (!updates.updatedFiles.empty()) {
+                        for (auto &ast : updates.updatedFileIndexes) {
+                            if (ast.tree) {
+                                updates.updatedFinalGSFileIndexes.emplace_back(std::move(ast));
```

Comment:
When we inline and constant fold commitFileUpdates into each function, is there anything holding us back from just writing into `this->indexedFinalGS` right here? i.e., could we delete this `updatedFinalGSFileIndexes` field from `LSPFileUpdates`?

Reply:
Nope, we could do that. We'll want to move the initialization of the UndoState earlier so that we're not modifying state before it's saved, but otherwise that would be a nice refactoring.